vault/token_store.go

```diff
@@ -917,11 +1083,155 @@ func (ts *TokenStore) destroyCubbyhole(saltedID string) error {
 
 func (ts *TokenStore) authRenew(
 	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	if req.Auth == nil {
+		return nil, fmt.Errorf("request auth is nil")
+	}
+
+	f := framework.LeaseExtend(req.Auth.Increment, 0, ts.System())
+
+	idInt, ok := req.Auth.InternalData["id"]
+	if !ok {
+		// Fall back here; this is pre-roles so there are no stored IDs, so use previous behavior
+		return f(req, d)
+	}
+
+	id, ok := idInt.(string)
+	if !ok {
+		return nil, fmt.Errorf("found id in internal data but could not interpret as string")
+	}
+
+	te, err := ts.Lookup(id)
+	if err != nil {
+		return nil, fmt.Errorf("error looking up token: %s", err)
+	}
+	if te == nil {
+		return nil, fmt.Errorf("no token entry found during lookup")
+	}
+
+	// No role? Use normal LeaseExtend semantics
+	if te.Role == "" {
+		return f(req, d)
+	}
+
+	role, err := ts.getTokenStoreRole(te.Role)
+	if err != nil {
+		return nil, fmt.Errorf("error looking up role %s: %s", te.Role, err)
+	}
+
+	if role == nil {
+		resp, err := f(req, d)
+		if resp != nil {
+			resp.AddWarning(fmt.Sprintf("The token was created via role %s, but that role could no longer be found. Renewal fell back to normal token renewal semantics; if this token was a periodic token, this could mean that it can no longer be renewed.", te.Role))
+		}
+		return resp, err
+	}
+
+	if role.Period != 0 {
+		req.Auth.TTL = role.Period
+		return &logical.Response{Auth: req.Auth}, nil
+	}
 
-	f := framework.LeaseExtend(0, 0, ts.System())
 	return f(req, d)
 }
 
+func (ts *TokenStore) getTokenStoreRole(name string) (*tsRoleEntry, error) {
```

Comment:
Choose to ignore this. I recently came across some documentation which said that we don't need to name the getter method starting with `get`. (tokenStoreRole, instead)